stdlib/public/runtime/DynamicCast.cpp

```diff
@@ -1514,41 +1514,49 @@ tryCastToClassExistentialViaSwiftValue(
   auto destExistentialLocation
     = reinterpret_cast<ClassExistentialContainer *>(destLocation);
 
-  // Fail if the target has constraints that make it unsuitable for
-  // a __SwiftValue box.
-  // FIXME: We should not have different checks here for
-  // Obj-C vs non-Obj-C.  The _SwiftValue boxes should conform
-  // to the exact same protocols on both platforms.
-  bool destIsConstrained = destExistentialType->NumProtocols != 0;
-  if (destIsConstrained) {
-#if SWIFT_OBJC_INTEROP // __SwiftValue is an Obj-C class
-    if (!findSwiftValueConformances(
-          destExistentialType, destExistentialLocation->getWitnessTables())) {
-      return DynamicCastResult::Failure;
-    }
-#else // __SwiftValue is a native class
-    if (!swift_swiftValueConformsTo(destType, destType)) {
-      return DynamicCastResult::Failure;
-    }
-#endif
-  }
+  switch (srcType->getKind()) {
+  case MetadataKind::Class:
+  case MetadataKind::ObjCClassWrapper:
+  case MetadataKind::ForeignClass:
+    // Class references always go directly into
+    // class existentials; it makes no sense to wrap them.
+    return DynamicCastResult::Failure;
 
+  /*
+  // I suspect metatypes should also never get boxed here,
+  // but the current test suite fails with this, so this needs
+  // more research.
```

Comment:
Is it failing on Darwin or just other platforms? Metatypes aren't objects on other platforms, so they'd have to be boxed. Seems like they could avoid boxing on Darwin, but maybe there's some subtlety there.

Reply:
Failing on Darwin, too, which is why I was so surprised.  It's possible this would be a bug in the test suite or even in the spec, but I've not yet had time to work through those particular test cases to figure out what's going on.  It's also possible that I'm just wrong and it's perfectly okay for metatypes to get boxed.  :shrug:

Reply:
Oh wait, it's only class metatypes that are objects on Darwin. The most accurate thing here would probably be to box everything on non-Darwin and box all non-class types on Darwin. That feels kind of weird though.

Reply:
Ah,  yes.  I remember at least one of the failing test was for a struct metatype.  That would explain it.  Best answer is probably to leave it as-is for now.

Reply:
Ack.  No, we have to avoid boxing class metatypes on Darwin.  Otherwise, we risk having those metatypes be unusable when we pass them to Obj-C code.  I'll fix that.

Reply:
Yeah, ObjC will get unhappy if you give it a SwiftValue wrapping a class. So, weird or not, seems like it has to be that way.

Reply:
I fleshed this out to avoid boxing class metatypes on Darwin.  And I also added a casting test case to verify that both struct metatypes and class metatypes can be round-tripped through Any, AnyObject, etc.